Add new Context parameter

MatthewDolan opened the pull request:
Potential solution to https://github.com/go-gorp/gorp/issues/344

MatthewDolan commented:
Couple of problems I just noticed with the build.

(1) The "context" package was added with Go 1.7 so adding it to the API would break compatibility with Go 1.6. Is that an issue? I guess I could copy it's definition into gorp, but that seems like it would break going forward.

(2) The Context API being added to db.Sql wasn't added until I think Go 1.8. I changed it so it will degrade gracefully to not passing on the context in earlier versions, but that will make the test I added fail unless I can exclude that test for earlier versions.

srenatus commented on gorp_test.go:
```diff
@@ -2360,6 +2361,31 @@ func TestPrepare(t *testing.T) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer time.AfterFunc(100 * time.Millisecond, cancel).Stop()
```
I suppose you did that on purpose, just wanted to point out that this is equivalent to:
```go
ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
defer cancel()
```

srenatus commented on gorp_test.go:
```diff
@@ -2360,6 +2361,31 @@ func TestPrepare(t *testing.T) {
 	}
 }
 
+func TestWithCanceledContext(t *testing.T) {
```
> (2) The Context API being added to db.Sql wasn't added until I think Go 1.8. [...]  but that will make the test I added fail unless I can exclude that test for earlier versions.

If you'd put that into its own file, you would be able to add `// +build !go1.7`, I believe? To make these tests not run on 1.7? (not entirely sure if it works like that, though)

nelsam commented:
!go1.7 would make it only run on 1.6 and earlier. it should be +build
go1.8.  Other than that, yes - anything that requires 1.8 should be gated
on build tags in separate files.

I think it's somewhere in the readme that we aim to support the two most
recent go versions. If I'm wrong about it being in the readme, well... I
still think supporting the two most recent versions is plenty.

On Aug 31, 2017 00:42, "Stephan Renatus" <notifications@github.com> wrote:

> *@srenatus* approved this pull request.
>
> Thanks for picking up that context issue!
>
> Looks good to me -- but I, of course, can't help with the open questions.
> No idea what compatibility gorp wants to provide.
> ------------------------------
>
> In gorp_test.go
> <https://github.com/go-gorp/gorp/pull/352#discussion_r136255399>:
>
> > @@ -2360,6 +2361,31 @@ func TestPrepare(t *testing.T) {
>  	}
>  }
>
> +func TestWithCanceledContext(t *testing.T) {
> +	if _, driver := dialectAndDriver(); driver != "mysql" {
> +		t.Skip("Cancellation is not yet supported by all drivers. Known to be supported by mysql")
> +	}
> +
> +	dbmap := initDbMap()
> +	defer dropAndClose(dbmap)
> +
> +	ctx, cancel := context.WithCancel(context.Background())
> +
> +	defer time.AfterFunc(100 * time.Millisecond, cancel).Stop()
>
> I suppose you did that on purpose, just wanted to point out that this is
> equivalent to:
>
> ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)defer cancel()
>
> ------------------------------
>
> In gorp_test.go
> <https://github.com/go-gorp/gorp/pull/352#discussion_r136256039>:
>
> > @@ -2360,6 +2361,31 @@ func TestPrepare(t *testing.T) {
>  	}
>  }
>
> +func TestWithCanceledContext(t *testing.T) {
>
> (2) The Context API being added to db.Sql wasn't added until I think Go
> 1.8. [...] but that will make the test I added fail unless I can exclude
> that test for earlier versions.
>
> If you'd put that into its own file, you would be able to add // +build
> !go1.7, I believe? To make these tests not run on 1.7? (not entirely sure
> if it works like that, though)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-gorp/gorp/pull/352#pullrequestreview-59766343>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA-QBGNy7lnyVAm_GE5lh3n1RYmLx43rks5sdlXMgaJpZM4PITIB>
> .
>

MatthewDolan commented:
Thanks for the comments @srenatus and @nelsam.

I removed 1.6 as a supported version for travis and added 1.8 and 1.9.

And I moved the cancelation test into a new file that only runs on 1.8 and above.

The tests are passing now for 1.7, 1.8, and 1.9.

bryanpaluch commented:
Hi @MatthewDolan I think you need to add support in a few more places. QueryRow -> QueryRowContext, Query -> QueryContext, Prepare -> PrepareContext, and pretty much any other time there is an alternative Context function call with *sql.DB.

MatthewDolan commented:
@bryanpaluch I think the decision was to do the other option in the discussion on https://github.com/go-gorp/gorp/issues/344. So instead of adding a new set of methods, you can chain `WithContext(ctx)` in.

i.e. replace `dbMap.Query(...)` with `dbMap.WithContext(ctx).Query(...)`.

Does that make sense or am I miss understanding your comment?

MatthewDolan commented:
@bryanpaluch Oh you are totally right. Let me take another look.

MatthewDolan commented:
@bryanpaluch Thanks. I updated the PR to add the other calls.

Build is passing for 1.7, 1.8, and 1.9.

nelsam commented:
I'll review this weekend.  I'm blocking out something on my schedule.

nelsam commented on gorp_go18_test.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+func TestWithCanceledContext(t *testing.T) {
+	if _, driver := dialectAndDriver(); driver != "mysql" {
```
I would really like to have a blacklist of drivers that *don't* support it, rather than a whitelist of drivers that *do*.  If you're unsure, assume there is support - if it's not failing in travis, we'll deal with it when it becomes a problem.

MatthewDolan replied:
Good call. It seems like mymysql is the only driver that has not added support (I filed it as an issue https://github.com/ziutek/mymysql/issues/138).

Ran into one other problem though. Some sql dialects (namely sqlite) don't support native sleep calls. I checked the test for the sqlite driver and it looks like they have an attempt at testing a canceled context (https://github.com/mattn/go-sqlite3/blob/master/sqlite3_go18_test.go#L113), but I don't think it's as clean as sleep. I just excluded the languages that don't support sleep for now.

MatthewDolan commented:
@nelsam Just in-case you missed my comment inline, I added an exclusion list instead of an inclusion list like you suggested.

Let me know if there is anything else I need to update.

Thanks for taking a look.

nelsam commented:
Yep, I'll check on it soon.  I've had a busy couple of days.

nelsam commented on gorp_go18_test.go:
```diff
@@ -0,0 +1,80 @@
+func TestWithCanceledContext(t *testing.T) {
+	dialect, driver := dialectAndDriver()
+	if unsupportedDrivers[driver] {
+		t.Skip("Cancellation is not yet supported by all drivers. Not known to be unsupported in %s.", driver)
```
I think this should read `Known to be unsupported` or `Not known to be supported`, rather than `Not known to be unsupported`.

nelsam commented on dialect.go:
```diff
@@ -49,6 +51,9 @@ type Dialect interface {
 	TruncateClause() string
 
+	// string to sleep for d duration
+	SleepClause(d time.Duration) string
```
This is a breaking change and will force a `v3` of `gorp`, since some `gorp` users may have implemented custom `Dialect` types.  Can we instead add a new interface type, say `SleepDialect`, such that:

```go
type SleepDialect interface {
    SleepClause(time.Duration) string
}
```

And then remove the `SleepClause` methods on dialects that don't have a sleep clause?

nelsam replied:
In addition, since this method is *only* used by tests, can we move the `SleepDialect` interface to the test package?  If it's in `gorp`'s main package, then users expect that implementing `SleepDialect` gives them something from `gorp`, rather than `gorp`'s tests.

nelsam commented on dialect_oracle.go:
```diff
@@ -93,6 +94,10 @@ func (d OracleDialect) TruncateClause() string {
 
+func (d OracleDialect) SleepClause(s time.Duration) string {
+	return ""
+}
```
If we were to use a separate `SleepDialect`, you could just skip implementing this method.

nelsam commented on dialect_sqlite.go:
```diff
@@ -88,6 +89,10 @@ func (d SqliteDialect) TruncateClause() string {
 
+func (d SqliteDialect) SleepClause(s time.Duration) string {
+	return ""
+}
```
Same here as with the OracleDialect - you wouldn't have to implement this method here.

nelsam commented on dialect_sqlserver.go:
```diff
@@ -103,6 +104,10 @@ func (d SqlServerDialect) TruncateClause() string {
 
+func (d SqlServerDialect) SleepClause(s time.Duration) string {
+	return ""
+}
```
Same as with Oracle and Sqlite

nelsam commented on gorp_go18_test.go:
```diff
@@ -0,0 +1,80 @@
+		t.Skip("Cancellation is not yet supported by all drivers. Not known to be unsupported in %s.", driver)
+	}
+
+	if dialect.SleepClause(time.Second) == "" {
```
If we used a `SleepDialect`, this would change to:

```go
if sleeper, ok := dialect.(gorp.SleepDialect); ok {
```

nelsam commented:
@MatthewDolan anything you want to add before merge? Otherwise, I can merge now.

MatthewDolan commented:
Everything looks good to me. Feel free to merge.

Thanks for looking through this.

nelsam commented:
Thanks for the submission!  I'm trying to be more active with `gorp`, but you know how life and work go.  Especially with `gorp` being something I no longer actively use.
